File: mathlib/derivations/simplifications/DivisionIdentity.ts
import { Argument } from "../../Argument";
import { num } from "../../ConvenientExpressions";
import { Expression } from "../../expressions/Expression";
import { Fraction } from "../../expressions/Fraction";
import { Integer } from "../../expressions/Integer";
import { Relationship } from "../../Relationship";
import { setOf } from "../../util/ThingsThatShouldBeInTheStdLib";
import { NoContextExpressionSimplificationRule } from "../NoContextExpressionSimplificationRule";

/**
 * Anything divided by 1 is the numerator
 */
export class DivisionIdentity extends NoContextExpressionSimplificationRule {
    protected appliesImpl(exp: Expression): boolean {
        return exp instanceof Fraction
            && exp.denominator instanceof Integer
            && exp.denominator.value == 1
    }
    protected applyImpl(exp: Expression): Set<Argument> {
        const frac = exp as Fraction
        return setOf(new Argument(setOf(frac), {
            n: frac,
            r: Relationship.Equal,
            n1: Fraction.of(frac.numerator, num(1)),
        }, "Division identity is 1"))
    }

}

File: ihateintegrals/src/derivation_rules/division_identity.rs
use super::DerivationRule;
use crate::argument::Argument;
use crate::expressions::Expression;
use std::rc::Rc;

/**
* x/1 = x
*/
pub struct DivisionIdentity {}

impl DerivationRule for DivisionIdentity {
    fn apply(&self, input: Expression) -> Vec<(Expression, Rc<Argument>)> {
        let fraction = match input {
            Expression::Fraction(ref f) => f,
            _ => return vec![],
        };

        if let Expression::Integer(integer) = fraction.denominator() {
            if integer.value() != 1 {
                return vec![];
            }
        } else {
            return vec![];
        }

        vec![(
            fraction.numerator(),
            Argument::new(String::from("Division identity"), vec![input], self.name()),
        )]
    }
    fn name(&self) -> String {
        String::from("DivisionIdentity")
    }
}

#[cfg(test)]
mod tests {
    use crate::{convenience_expressions::i, expressions::Fraction};

    use super::DerivationRule;
    use super::DivisionIdentity;

    #[test]
    fn test_1() {
        let start = Fraction::of(i(2), i(1));
        let rule = DivisionIdentity {};
        let result = rule.apply(start);
        assert_eq!(result.first().unwrap().0, i(2));
    }
}
